src/types.ts

```diff
@@ -89,3 +92,68 @@ export interface AmplitudeReactNativeModule {
     eventUploadThreshold: number,
   ): Promise<boolean>;
 }
+
+/**
+ * Strings that have special meaning when used as an event's type
+ * and have different specifications.
+ */
+export enum SpecialEventType {
+  IDENTIFY = '$identify',
+  GROUP_IDENTIFY = '$groupidentify',
+}
+
+export type BaseEvent = {
+  eventType: Exclude<string, SpecialEventType>;
+  userId?: string;
+  deviceId?: string;
+  eventProperties?: PropertiesObject;
+};
+
+export type IdentifyEvent = {
+  eventType: SpecialEventType.IDENTIFY;
+  userId?: string;
+  deviceId?: string;
+  userProperties: IdentifyPayload;
+};
+
+export type GroupIdentifyEvent = {
+  eventType: SpecialEventType.GROUP_IDENTIFY;
+  userId?: string;
+  deviceId?: string;
+  groupType: string;
+  groupName: string | string[];
+  groupProperties: IdentifyPayload;
+};
+
+export type Event = BaseEvent | IdentifyEvent | GroupIdentifyEvent;
+
+/**
+ * Unstructured object to let users pass extra data to middleware
+ */
+export interface MiddlewareExtra {
+  [name: string]: any;
+}
+
+/**
+ * Data to be processed by middleware
+ */
+export interface MiddlewarePayload {
+  event: Event;
+  extra?: MiddlewareExtra;
+}
+
+/**
+ * Function called at the end of each Middleware to run the next middleware in the chain
+ */
+export type MiddlewareNext = (payload: MiddlewarePayload) => void;
+
+/**
+ * A function to run on the Event stream (each logEvent call)
+ *
+ * @param payload The event and extra data being sent
+ * @param next Function to run the next middleware in the chain, not calling next will end the middleware chain
+ */
+export type Middleware = (
```

Comment:
Can you please add an example of middleware usage in the `example` project.

```
  amplitudeInstance.addEventMiddleware((payload, next) => {
    const { event, extra } = payload;
    console.log(
      `$event=${event.eventType} properties=${event.eventProperties} extra=${extra}`,
    );
    next(payload);
  });

```

Reply:
Added